<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>rwrap: Seamlessly integrate R packages into Python &mdash; kpj&#39;s blog 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=d45e8c67"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Vydia: A modularized video player" href="vydia.html" />
    <link rel="prev" title="pyskim: Quick summary statistics for dataframes" href="pyskim.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            kpj's blog
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Linux</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../linux/ConfigureFreshArchlinuxInstall.html">Configuring a fresh Archlinux Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/Eduroam.html">Connecting to Eduroam Network</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/EncryptPartitions.html">Encrypting Partitions using LUKS</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/KernelBuilding.html">Building a Custom Linux Kernel</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/MediaServerSetup.html">Setting up a media server</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/MonitoringWithMunin.html">Monitoring server resources using Munin</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/RescuingABrokenSystem.html">Rescuing a Broken Archlinux System</a></li>
<li class="toctree-l1"><a class="reference internal" href="../linux/SetupEncryptedArchlinux.html">Setting up Archlinux with Encrypted Disk</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data Mining &amp; Visualization</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/ArtOfLife.html">The Art of Life</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/Eurostat_Visualizations.html">Visualizing Eurostat data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/IMDB_ratings.html">Investigating TV Series ratings using IMDB</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/MEP_Statistics.html">Analyzing the European Parliament</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/ProjectWordcloud.html">Creating a Wordcloud of Your (Work) Life</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/SortingAnimations.html">Visualizing Sorting Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mining_visualization/SwissMountainProfiles.html">Visualizing Height Profiles of Swiss Mountains</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Statistics &amp; Machine Learning</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../stats_ml/Autoencoder.html">Playing with Variational Autoencoders</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats_ml/LinearRegressionCoefficients.html">Interpreting the coefficients of a linear regression</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats_ml/ProbabilityDistributionsOverview.html">The Daring Landscape of Probability Distribution Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../misc/BenchmarkingCommonOperations.html">Benchmarking common operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/InterceptingHTTPTraffic.html">Intercepting HTTP(S) Traffic of Android Applications</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/PythonAlmanac.html">The Python Almanac</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/ThreadingVSMultiprocessing.html">Comparing threading and multiprocessing</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">OSS Projects</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="bin2fasta.html">bin2fasta: Store any file as a fasta file</a></li>
<li class="toctree-l1"><a class="reference internal" href="clypper.html">clypper: Rapidly create supercuts from various video sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="cookiecutter-python.html">cookiecutter-python: An opinionated template for Python packages</a></li>
<li class="toctree-l1"><a class="reference internal" href="gitviz.html">GitViz: Visualize the evolution of a git repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="ipy_pdcache.html">ipy_pdcache: Automatically cache results of intensive computations in IPython</a></li>
<li class="toctree-l1"><a class="reference internal" href="jupyfmt.html">jupyfmt: The uncompromising Jupyter notebook formatter</a></li>
<li class="toctree-l1"><a class="reference internal" href="metagenompy.html">metagenompy: The NCBI taxonomy as a NetworkX graph</a></li>
<li class="toctree-l1"><a class="reference internal" href="pdftty.html">pdftty: A PDF viewer for the terminal</a></li>
<li class="toctree-l1"><a class="reference internal" href="plyder.html">plyder: Download Manager with Web-Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="pyskim.html">pyskim: Quick summary statistics for dataframes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">rwrap: Seamlessly integrate R packages into Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="vydia.html">Vydia: A modularized video player</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Scientific Work</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../science/presentations.html">Presentations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../science/publications.html">Publications</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">kpj's blog</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">rwrap: Seamlessly integrate R packages into Python</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/my_projects/rwrap.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rwrap-seamlessly-integrate-r-packages-into-python">
<h1>rwrap: Seamlessly integrate R packages into Python<a class="headerlink" href="#rwrap-seamlessly-integrate-r-packages-into-python" title="Link to this heading">ÔÉÅ</a></h1>
<p>The world of data science is largely divided into the Python pioneers and R rascals (in a loving way) with a few honorable mentions.
These two camps strive for supremacy, each offering their own set of distinctive advantages. This shall not be a review of these, but simply highlight that the <a class="reference external" href="https://www.tidyverse.org/">tidyverse</a> and <a class="reference external" href="https://www.bioconductor.org/">Bioconductor</a> ecosystems make a strong case for R.
However, since Python is evidently the better language (citation needed), accessing the vast amount of R-specific functionality offered by Bioconductor packages directly from Python would be quite glamorous.</p>
<p><a class="reference external" href="https://github.com/kpj/rwrap">rwrap</a> aims at doing exactly that. By providing a wrapper around <a class="reference external" href="https://rpy2.github.io/">rpy2</a> and adding many additional data conversion rules, it removes the need for loads of boilerplate code and makes using R packages in Python easier.</p>
<p>For example, running a Differential Gene Expression analysis and adding genomic annotations using the R packages <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a> and <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> can now look like this in Python:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">rwrap</span> <span class="kn">import</span> <span class="n">DESeq2</span><span class="p">,</span> <span class="n">biomaRt</span><span class="p">,</span> <span class="n">base</span><span class="p">,</span> <span class="n">stats</span>


<span class="n">DESeq2</span>
<span class="c1">## &lt;module &#39;DESeq2&#39; from &#39;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/DESeq2&#39;&gt;</span>
<span class="n">biomaRt</span>
<span class="c1">## &lt;module &#39;biomaRt&#39; from &#39;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/biomaRt&#39;&gt;</span>


<span class="c1"># retrieve count data (https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP009615)</span>
<span class="n">df_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s2">&quot;http://duffel.rail.bio/recount/v2/SRP009615/counts_gene.tsv.gz&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
<span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;gene_id&quot;</span><span class="p">)</span>
<span class="n">df_design</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span><span class="s2">&quot;condition&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;6&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;6&quot;</span><span class="p">]}</span>
<span class="p">)</span>

<span class="c1"># run differential gene expression analysis</span>
<span class="n">dds</span> <span class="o">=</span> <span class="n">DESeq2</span><span class="o">.</span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span>
    <span class="n">countData</span><span class="o">=</span><span class="n">df_counts</span><span class="p">,</span> <span class="n">colData</span><span class="o">=</span><span class="n">df_design</span><span class="p">,</span> <span class="n">design</span><span class="o">=</span><span class="n">stats</span><span class="o">.</span><span class="n">as_formula</span><span class="p">(</span><span class="s2">&quot;~ condition&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">dds</span> <span class="o">=</span> <span class="n">DESeq2</span><span class="o">.</span><span class="n">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>

<span class="n">res</span> <span class="o">=</span> <span class="n">DESeq2</span><span class="o">.</span><span class="n">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">))</span>
<span class="n">df_res</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">as_data_frame</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

<span class="c1"># annotate result</span>
<span class="n">ensembl</span> <span class="o">=</span> <span class="n">biomaRt</span><span class="o">.</span><span class="n">useEnsembl</span><span class="p">(</span><span class="n">biomart</span><span class="o">=</span><span class="s2">&quot;genes&quot;</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s2">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
<span class="n">df_anno</span> <span class="o">=</span> <span class="n">biomaRt</span><span class="o">.</span><span class="n">getBM</span><span class="p">(</span>
    <span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_id_version&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_biotype&quot;</span><span class="p">],</span>
    <span class="n">filters</span><span class="o">=</span><span class="s2">&quot;ensembl_gene_id_version&quot;</span><span class="p">,</span>
    <span class="n">values</span><span class="o">=</span><span class="n">df_res</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
    <span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ensembl_gene_id_version&quot;</span><span class="p">)</span>

<span class="n">df_res</span> <span class="o">=</span> <span class="n">df_res</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_anno</span><span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;padj&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_res</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>  <span class="c1"># pd.DataFrame</span>
<span class="c1">##                      baseMean  log2FoldChange     lfcSE      stat        pvalue          padj          gene_biotype</span>
<span class="c1">## ENSG00000222806.1  158.010377       22.137400  2.745822  8.062214  7.492501e-16  2.853744e-11       rRNA_pseudogene</span>
<span class="c1">## ENSG00000255099.1   65.879611       21.835651  2.915452  7.489627  6.906949e-14  1.315359e-09  processed_pseudogene</span>
<span class="c1">## ENSG00000261065.1   92.351998       22.273400  3.144991  7.082182  1.419019e-12  1.351190e-08                lncRNA</span>
<span class="c1">## ENSG00000249923.1  154.037908       18.364027  2.636083  6.966407  3.251381e-12  2.476772e-08                lncRNA</span>
<span class="c1">## ENSG00000267658.1   64.371181      -19.545702  3.041247 -6.426871  1.302573e-10  8.268736e-07                lncRNA</span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pyskim.html" class="btn btn-neutral float-left" title="pyskim: Quick summary statistics for dataframes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="vydia.html" class="btn btn-neutral float-right" title="Vydia: A modularized video player" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, kpj.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>